[BotService] Updates, enhancements and cleanup in az bot

stevengum opened the pull request:
- Remove extraneous `--version` flag for `az bot create -k webapp` (Only v4 bots are supported) (Fixes: microsoft/botframework-sdk#5669)
- Add support for updating the Icon for a bot (Fixes: microsoft/botbuilder-tools#1198)
- Add name availability check to `az bot create` (Fixes: microsoft/botbuilder-tools#1287)
- Add `--enable-enhanced-auth` flag to `az bot directline create`
- Add `az bot directline update` (Fixes: Azure/azure-cli#9954)
- Remove preview tag from some of the commands & command groups (Fixes: microsoft/botframework-sdk#5667)
- Loosen `--password` requirement for `az bot create` -  it is only required for creating Web App Bots, not Registration Bots
- Factor out `az bot prepare-deploy` tests into `BotLocalErrorsTests` using `(unittest.TestCase)` as these tests deal with local file creation for deploying bots to Azure.
- Fix `az bot prepare-deploy` changing `--proj-file-path` value to lower case (e.g. `"Test.csproj"` to `"test.csproj"`) (Fixes: #11390)

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [X] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

fengzhou-msft commented:
@Juliehzl please review this PR when it's ready.

stevengum commented:
@Juliehzl this PR is ready for review 🚀

yonzhan commented:
@Juliehzl @jsntcy  please help with the PR.

Juliehzl commented on src/azure-cli/HISTORY.rst:
```diff
@@ -56,6 +56,15 @@ Release History
 
 * Fix issue in az backup policy list-associated-items. Added optional BackupManagementType parameter.
 
+**BotService**
```
Please rebase dev branch and make your release note available for next release.

stevengum replied:
[Done](https://github.com/Azure/azure-cli/pull/11318/files#diff-364518622fafca75168eac23785eaca6R3-R14)!

stevengum commented:
@Juliehzl in addition to addressing your feedback, I added another bug fix for Azure/azure-cli#11390 via commit https://github.com/Azure/azure-cli/pull/11318/commits/2ae16ed5cf76bcd29dd7c0d4955ec3110807e27b.

jsntcy commented on src/azure-cli/azure/cli/command_modules/botservice/channel_operations.py:
```diff
@@ -19,6 +23,25 @@ def create_channel(client, channel, channel_name, resource_group_name, resource_
+    return client.update(
+        resource_group_name=resource_group_name,
+        resource_name=resource_name,
+        channel_name='DirectLineChannel',
```
Use channel_name=channel_name instead of channel_name='DirectLineChannel' to make update_channel to be a more generic method?

stevengum replied:
We don't currently support updating other channels, but I've applied this feedback.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -60,10 +60,9 @@ def load_arguments(self, _):
         c.argument('tags', arg_type=tags_type)
-        c.argument('language', options_list=['--lang'], arg_type=get_enum_type(SUPPORTED_LANGUAGES), help='The language to be used to create the bot.', arg_group='Web Bot Specific')
-        c.argument('version', options_list=['-v', '--version'], help='The Microsoft Bot Builder SDK version to be used to create the bot', arg_type=get_enum_type(['v4']), arg_group='Web Bot Specific')
+        c.argument('language', options_list=['--lang'], arg_type=get_enum_type(SUPPORTED_LANGUAGES), help='The language to be used to create the bot.', arg_group='Web App Bot Specific')
```
Could you please use options_list=['--language', '-l']?

stevengum replied:
`'-l'` would conflict with the global `'--location', '-l'` I think...

I could update it to be `options_list=['--language', '--lang']`, or should I leave it as is?

stevengum replied:
After checking the code, `-l` would conflict with the `location` parameter

Juliehzl replied:
It will be better to provide one full parameter name for language. What about ['--language', '-a']? ['--language','---lang'] is also acceptable.

jsntcy replied:
I prefer  ['--language','---lang'] as 'a' is not intuit.

stevengum replied:
@Juliehzl and @jsntcy I'm open to `['--language', '--lang']`, is this acceptable?

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -60,10 +60,9 @@ def load_arguments(self, _):
         c.argument('endpoint', options_list=['-e', '--endpoint'], help='The messaging endpoint of the bot.', arg_group='Registration Bot Specific')
         c.argument('msa_app_id', options_list=['--appid'], help='The Microsoft account ID (MSA ID) to be used with the bot.')
-        c.argument('password', options_list=['-p', '--password'], help='The Microsoft account (MSA) password for the bot. Used to authorize messages being sent to the bot.')
+        c.argument('password', options_list=['-p', '--password'], help='The Microsoft account (MSA) password for the bot. Used to authorize messages being sent to the bot. Required for Web App Bots.', arg_group='Web App Bot Specific')
```
What do you want to show for your help message with "arg_group='Web App Bot Specific"? Could you make it more clear?

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/_params.py:
```diff
@@ -163,6 +163,16 @@ def load_arguments(self, _):
         c.argument('site_name', options_list=['-s', '--site-name'], help='Name of the Directline channel site.')
         c.argument('is_v1_disabled', options_list=['--disablev1'], help='If true, v1 protocol will be disabled on the channel', arg_type=get_three_state_flag())
         c.argument('is_v3_disabled', options_list=['--disablev3'], help='If true, v3 protocol will be disabled on the channel.', arg_type=get_three_state_flag())
+        c.argument('is_secure_site_enabled', help='If true, enables enhanced authentication features. Must be true for --trusted-origins parameters to work.', arg_type=get_three_state_flag())
```
Have you placed these new added arguments for `az bot directline create` in release notes?

Juliehzl replied:
Typically, we will recommend use something like "--enable-xxx" as argument in CLI.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/channel_operations.py:
```diff
@@ -19,6 +23,25 @@ def create_channel(client, channel, channel_name, resource_group_name, resource_
     )
 
 
+def update_channel(client, channel, channel_name, resource_group_name, resource_name):
```
One more question here is that channel_name is required for you function. But where do you use channel_name here?

stevengum replied:
I've remedied this by addressing @jsntcy's [feedback](https://github.com/Azure/azure-cli/pull/11318#discussion_r352514008).

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/custom.py:
```diff
@@ -74,15 +75,27 @@ def __prepare_configuration_file(cmd, resource_group_name, kudu_client, folder_p
     bot_kind = 'bot'
     webapp_kind = 'webapp'
 
+    # Mapping: registration is deprecated, we now use 'bot' kind for registration bots
```
When do you deprecate registration bots? Can we show up warning message here to inform users such change?

stevengum replied:
We haven't deprecated the "Bot Channels Registration" Azure Resource, the `kind` was changed over a year ago, and to preserve the current flow for users we just mapped `'registration'` to the correct value `'bot'`. 

This new line isn't new code, it's just restructured around the new name availability check. ([link to original location](https://github.com/Azure/azure-cli/pull/11318/files#diff-29e4a6bde35880c6917ba0382e947063L108-L111)).

I think `'registration'` is a clearer indicator of what the user is signing up for in regards to the Azure Bot Service, as opposed to `'bot'`, which is redundant and not clear.

If it is required, I can change the enum values to `'bot'` and `'webapp'`, in the [params](https://github.com/Azure/azure-cli/pull/11318/files#diff-9f85bf3488eb6b9381fd84553e8ef64aR58) and start the deprecation process for the `'registration'` value.

```py
c.argument('kind', options_list=['--kind', '-k'], arg_type=get_enum_type(['registration', 'webapp']), help='The kind of the bot.')
```

Juliehzl commented on src/azure-cli/azure/cli/command_modules/botservice/custom.py:
```diff
@@ -74,15 +75,27 @@ def __prepare_configuration_file(cmd, resource_group_name, kudu_client, folder_p
+    # Check the resource name availability for the bot.
+    name_response = NameAvailability.check_name_availability(client, resource_name, kind)
+    if not name_response.valid:
+        # If the name is unavailable, gracefully exit and log the reason for the user.
```
Should we throw exception here?

stevengum replied:
That makes much more sense, I'll change the code to do that instead.

Juliehzl commented:
@stevengum Please resolve the left comments and make CI pass, then I will approve.

Juliehzl commented:
Hi @@stevengum, please resolve conflicts.

stevengum commented:
@Juliehzl I've rebased the branch and resolved the merge conflicts in HISTORY.rst
